include/linux/file_compat.h

```diff
@@ -76,8 +76,17 @@ spl_filp_fallocate(struct file *fp, int mode, loff_t offset, loff_t len)
 #define	spl_filp_fsync(fp, sync)	vfs_fsync(fp, (fp)->f_dentry, sync)
 #endif /* HAVE_2ARGS_VFS_FSYNC */
 
+#ifdef HAVE_INODE_LOCK_SHARED
+#define	spl_inode_lock(ip)		inode_lock_nested(ip, I_MUTEX_PARENT)
+#define	spl_inode_unlock(ip)		inode_unlock(ip)
+#define	spl_inode_lock_shared(ip)	inode_lock_shared(ip)
+#define	spl_inode_unlock_shared(ip)	inode_unlock_shared(ip)
+#else
 #define	spl_inode_lock(ip)		mutex_lock_nested(&(ip)->i_mutex, I_MUTEX_PARENT)
 #define	spl_inode_unlock(ip)		mutex_unlock(&(ip)->i_mutex)
+#define	spl_inode_lock_shared(ip)	mutex_lock(&(ip)->i_mutex)
```

Comment:
The shared version should also use `mutex_lock_nested()` and pass I_MUTEX_PARENT for the reason listed in 628fc52137fc14377eba46c66b57d8d094e88507.  This way we won't get lockdep warnings on older kernels running newer versions of ZFS which use `spl_inode_lock_shared()`.